### PR TITLE
remove implicit restore incorrect optimization

### PR DESCRIPTION
**Pull request by OsmondJiang**
#4434

### CODE REVIEW

**Comment by OsmondJiang** [comment]
We will hold on this PR until we find a easy way to:
1. Remove implicit wrong logic
2. Keep the perf as same as right now.

I am going to add "--no-restore" to all our docfx callers

**Comment by OsmondJiang** [comment]
The main perf concern is here:

image


Every test now will fetch remote, no matter the test dependencies repo exists or not.

cc @live1206  @yufeih

File: src/docfx/cli/Program.cs [hunk 1]
```diff
@@ -71,12 +71,14 @@ internal static async Task<int> Run(string[] args)
                             Done(command, stopwatch.Elapsed, report);
                             break;
                         case "build":
-                            await Restore.Run(docset, options, report, @implicit: true);
+                            if (!options.NoRestore)
```
Comment (yufeih) [inline src/docfx/cli/Program.cs]: remove `--no-restore`

Comment (OsmondJiang) [reply src/docfx/cli/Program.cs]: i will do it in another PR

**Comment by herohua** [comment]
what the error message will be like if someone runs `build` without explicitly `restore` first?

**Comment by OsmondJiang** [comment]
> what the error message will be like if someone runs `build` without explicitly `restore` first?

In this PR, no error will thrown because docfx will automatically run `restore` before build. this PR just removed some incorrectly optimization for restore.

In my next PR, we will totally remove `implicit restore` feature, use will get `need-restore` error if some dependency repositories are not restored before build like below:

image

File: src/docfx/restore/RestoreFile.cs [hunk 2]
```diff
@@ -13,21 +13,18 @@ namespace Microsoft.Docs.Build
...
+                restoreUrl => Restore(restoreUrl, config));
         }
 
-        public static async Task Restore(string url, Config config, bool @implicit = false)
+        public static async Task Restore(string url, Config config)
         {
             var filePath = GetRestoreContentPath(url);
 
-            var (existingContent, existingEtagContent) = RestoreMap.TryGetRestoredFileContent(url);
-            if (!string.IsNullOrEmpty(existingContent) && @implicit)
```
Comment (live1206) [inline src/docfx/restore/RestoreFile.cs]: @OsmondJiang Just realized, we always need to restore from remote even the content already exists?

Comment (OsmondJiang) [reply src/docfx/restore/RestoreFile.cs]: no, we don't

**Pull request merged**